File: ios/oplayer/bitshares/BinSerializer.m
//
//  BinSerializer.m
//
//  Created by SYALON on 13-9-3.
//
//

#import "BinSerializer.h"
#import "ChainObjectManager.h"
#import "GraphenePublicKey.h"
#include "varint.h"
#include "bts_wallet_core.h"

/**
 *  varint编码的时候临时用到的缓冲区
 */
static char _varint_tmpbuf[10];

@interface BinSerializer()
{
    NSMutableData*  _data;
    NSRange         _range;
}
@end

@implementation BinSerializer

- (id)initForReader:(NSData*)data
{
    self = [super init];
    if (self)
    {
        //  TODO:6.0 没必要copy
        _data = [data mutableCopy];
        _range = NSMakeRange(0, 0);
    }
    return self;
}

- (id)initForWriter
{
    self = [super init];
    if (self)
    {
        _data = [NSMutableData data];
        _range = NSMakeRange(0, 0);
    }
    return self;
}

- (uint8_t)read_u8
{
    uint8_t value;
    _range.length = sizeof(value);
    [_data getBytes:&value range:_range];
    _range.location += _range.length;
    return value;
}

- (uint16_t)read_u16
{
    uint16_t value;
    _range.length = sizeof(value);
    [_data getBytes:&value range:_range];
    _range.location += _range.length;
    return value;
}

- (uint32_t)read_u32
{
    uint32_t value;
    _range.length = sizeof(value);
    [_data getBytes:&value range:_range];
    _range.location += _range.length;
    return value;
}

- (uint64_t)read_u64
{
    uint64_t value;
    _range.length = sizeof(value);
    [_data getBytes:&value range:_range];
    _range.location += _range.length;
    return value;
}

- (int64_t)read_s64
{
    int64_t value;
    _range.length = sizeof(value);
    [_data getBytes:&value range:_range];
    _range.location += _range.length;
    return value;
}

- (uint32_t)read_varint32
{
    char* all_bytes = (char*)[_data bytes];
    
    //  读取
    char* buf = (char*)&all_bytes[_range.location];
    int len = (int)([_data length] - _range.location);
    unsigned char size = 0;
    unsigned long long value = varint_decode(buf, len, &size);
    
    //  递增
    _range.location += (NSUInteger)size;
    
    //  返回
    return (uint32_t)value;
}

- (NSString*)read_object_id:(EBitsharesObjectType)object_type
{
    uint32_t value = [self read_varint32];
    return [NSString stringWithFormat:@"1.%@.%@", @(object_type), @(value)];
}

- (NSData*)read_bytes:(uint32_t)size
{
    if (size == 0){
        size = [self read_varint32];
    }
    if (size > 0){
        _range.length = size;
        id d = [_data subdataWithRange:_range];
        _range.location += _range.length;
        return d;
    } else {
        return [NSData data];
    }
}

- (NSString*)read_string
{
    return [[NSString alloc] initWithData:[self read_bytes:0] encoding:NSUTF8StringEncoding];
}

- (NSString*)read_public_key
{
    secp256k1_pubkey_compressed pubkey = {0, };
    
    _range.length = sizeof(pubkey.data);
    [_data getBytes:pubkey.data range:_range];
    _range.location += _range.length;
    
    return [[[GraphenePublicKey alloc] initWithSecp256k1PublicKey:&pubkey] toWifString];
}

- (BinSerializer*)write_u8:(uint8_t)value
{
    [_data appendBytes:&value length:sizeof(value)];
    return self;
}

- (BinSerializer*)write_u16:(uint16_t)value
{
    [_data appendBytes:&value length:sizeof(value)];
    return self;
}

- (BinSerializer*)write_u32:(uint32_t)value
{
    [_data appendBytes:&value length:sizeof(value)];
    return self;
}

- (BinSerializer*)write_u64:(uint64_t)value
{
    [_data appendBytes:&value length:sizeof(value)];
    return self;
}

- (BinSerializer*)write_s64:(int64_t)value
{
    [_data appendBytes:&value length:sizeof(value)];
    return self;
}

- (BinSerializer*)write_varint32:(uint64_t)value
{
    unsigned char size = 0;
    char* buf = varint_encode((unsigned long long)value, _varint_tmpbuf, sizeof(_varint_tmpbuf), &size);
    [_data appendBytes:buf length:(NSUInteger)size];
    return self;
}

/**
 *  写入对象ID类型（格式：x.x.x）
 */
- (BinSerializer*)write_object_id:(NSString*)x_x_x object_type:(EBitsharesObjectType)object_type
{
    //  convert 1.x.n into just n
    NSString* object_id_regular = @"^[0-9]+.[0-9]+.[0-9]+$";
    NSPredicate* object_id_pre = [NSPredicate predicateWithFormat:@"SELF MATCHES %@", object_id_regular];
    if ([object_id_pre evaluateWithObject:x_x_x])
    {
        //  TODO:integerValue是否溢出
        id ary = [x_x_x componentsSeparatedByString:@"."];
        NSAssert([[ary objectAtIndex:1] integerValue] == object_type, @"Invalid object id.");
        id object_id = [ary objectAtIndex:2];
        [self write_varint32:(uint64_t)[object_id integerValue]];
    }else{
        //  TODO:格式不匹配应该属于异常了
        [self write_varint32:(uint64_t)[x_x_x integerValue]];   //  TODO:integerValue是否益处
    }
    
    return self;
}

/**
 *  写入公钥对象。REMARK：BTS开头的地址。
 */
- (BinSerializer*)write_public_key:(NSString*)public_key_address
{
    secp256k1_pubkey_compressed pubkey = {0, };
    bool ret = __bts_gen_public_key_from_b58address((const unsigned char*)[public_key_address UTF8String], (const size_t)[public_key_address length],
                                                    [[ChainObjectManager sharedChainObjectManager].grapheneAddressPrefix length], &pubkey);
    if (!ret){
        //  TODO:无效公钥地址，不写入。
        return self;
    }
    
    //  写入33字节压缩公钥
    [_data appendBytes:pubkey.data length:sizeof(pubkey.data)];
    
    return self;
}

- (BinSerializer*)write_string:(NSString*)str
{
    [self write_bytes:[str dataUsingEncoding:NSUTF8StringEncoding] with_size:YES];
    return self;
}

- (BinSerializer*)write_fix_string:(NSString*)str
{
    [_data appendBytes:[str UTF8String] length:str.length];
    return self;
}

- (BinSerializer*)write_bytes:(NSData*)data with_size:(BOOL)with_size
{
    NSUInteger size = data.length;
    if (with_size){
        [self write_varint32:(uint64_t)size];
    }
    if (size > 0){
        [_data appendData:data];
    }
    return self;
}

- (NSData*)getData
{
    return [_data copy];
}

@end

File: android/app/src/main/java/bitshares/BinSerializer.kt
package bitshares

import VarInt.VarInt
import com.btsplusplus.fowallet.NativeInterface
import com.fowallet.walletcore.bts.ChainObjectManager
import com.macfaq.io.LittleEndianOutputStream
import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder

class BinSerializer {

    var _data: ByteArrayOutputStream = ByteArrayOutputStream()
    var _io: LittleEndianOutputStream = LittleEndianOutputStream(_data)

    var _reader: ByteBuffer? = null

    fun initForReader(data: ByteArray): BinSerializer {
        _reader = ByteBuffer.wrap(data)
        _reader!!.order(ByteOrder.LITTLE_ENDIAN)
        return this
    }

    fun read_u8(): Int {
        return _reader!!.get().toInt()
    }

    fun read_u16(): Int {
        return _reader!!.short.toInt()
    }

    fun read_u32(): Long {
        return _reader!!.int.toLong()
    }

    fun read_u64(): Long {
        //  TODO:Long < u64
        return _reader!!.long
    }

    fun read_s64(): Long {
        return _reader!!.long
    }

    fun read_varint32(): Int {
        return VarInt.getVarInt(_reader!!)
    }

    fun read_object_id(object_type: EBitsharesObjectType): String {
        val value = read_varint32()
        return "1.${object_type.value}.$value"
    }

    fun read_bytes(size: Int): ByteArray {
        val len = if (size == 0) read_varint32() else size
        if (len > 0) {
            val dst = ByteArray(len)
            _reader!!.get(dst, 0, len)
            return dst
        } else {
            return ByteArray(0)
        }
    }

    fun read_string(): String {
        return read_bytes(0).utf8String()
    }

    fun read_public_key(): String {
        return GraphenePublicKey().initWithSecp256k1PublicKey(read_bytes(33)).toWifString()
    }

    fun get_data(): ByteArray {
        return _data.toByteArray()
    }

    fun write_u8(value: Int): BinSerializer {
        _io.writeByte(value)
        return this
    }

    fun write_u16(value: Int): BinSerializer {
        _io.writeShort(value)
        return this
    }

    fun write_u32(value: Long): BinSerializer {
        _io.writeInt(value.toInt())
        return this
    }

    fun write_u64(value: Long): BinSerializer {
        //  TODO:Long < u64
        _io.writeLong(value)
        return this
    }

    fun write_s64(value: Long): BinSerializer {
        _io.writeLong(value)
        return this
    }

    fun write_varint32(value: Int): BinSerializer {
        val varint_tmpbuf = ByteArray(10)
        val offset = VarInt.putVarInt(value, varint_tmpbuf, 0)
        _io.write(varint_tmpbuf, 0, offset)
        return this
    }

    /**
     * 写入对象ID类型（格式：x.x.x）
     */
    fun write_object_id(value: String, object_type: EBitsharesObjectType): BinSerializer {
        //  TODO:格式验证 x.x.xxx
        val ary = value.split(".")
        assert(ary[1].toInt() == object_type.value) { "Invalid object id." }
        val oid = ary[ary.size - 1]
        this.write_varint32(oid.toInt())
        return this
    }

    /**
     * 写入公钥对象。REMARK：BTS开头的地址。
     */
    fun write_public_key(public_key_address: String): BinSerializer {
        val public_key = NativeInterface.sharedNativeInterface().bts_gen_public_key_from_b58address(public_key_address.utf8String(), ChainObjectManager.sharedChainObjectManager().grapheneAddressPrefix.utf8String())
        if (public_key == null) {
            //  无效公钥地址，不写入。
            return this
        }
        //  写入33字节压缩公钥
        _io.write(public_key)
        return this
    }

    /**
     * 写入字符串（变长，包含长度信息）
     */
    fun write_string(value: String): BinSerializer {
        write_bytes(value.utf8String(), with_size = true)
        return this
    }

    /**
     * 写入字符串（定长，不包含长度信息）
     */
    fun write_fix_string(value: String): BinSerializer {
        _io.write(value.utf8String())
        return this
    }

    /**
     * 写入2进制流
     */
    fun write_bytes(value: ByteArray, with_size: Boolean): BinSerializer {
        if (with_size) {
            this.write_varint32(value.size)
        }
        _io.write(value)
        return this
    }
}